src/util/rng.cpp

```diff
@@ -0,0 +1,62 @@
+/*********************                                                        */
+/*! \file rng.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief A Random Number Generator.
+ **
+ ** A random number generator, implements the xorshift* generator
+ ** (see S. Vigna, An experimental exploration of Marsaglia's xorshift
+ ** generators, scrambled. ACM Trans. Math. Softw. 42(4): 30:1-30:23, 2016).
+ **/
+
+#include "util/rng.h"
+
+#include <cfloat>
+#include "base/cvc4_assert.h"
+
+namespace CVC4 {
+
+uint64_t RNG::rand()
+{
+  /* xorshift* generator (see S. Vigna, An experimental exploration of
+   * Marsaglia's xorshift generators, scrambled. ACM Trans. Math. Softw.
+   * 42(4): 30:1-30:23, 2016). */
+  d_state ^= d_state >> 12;
+  d_state ^= d_state << 25;
+  d_state ^= d_state >> 27;
+  d_state *= UINT64_C(2685821657736338717);
+  return d_state;
+}
+
+uint64_t RNG::pick(uint64_t from, uint64_t to)
+{
+  Assert(from <= to);
+  Assert(to <= UINT64_MAX);
+  from = from == UINT64_MAX ? UINT64_MAX - 1 : from;
+  to = to == UINT64_MAX ? UINT64_MAX - 1 : to;
+  return (RNG::rand() % (to - from + 1)) + from;
+}
+
+double RNG::pickDouble(double from, double to)
+{
+  Assert(from <= to);
+  Assert(to <= DBL_MAX);
+  return RNG::rand() * (to - from ) + from;
+}
+
+bool RNG::pickWithProb(double probability)
+{
+  Assert(probability <= 1);
+  uint64_t p = (uint64_t) (probability * 1000);
+  uint64_t r = pick(0, 1000);
+  return r < p;
```

Comment:
There is a minor problem: If `probability = 1.0`, then `r` could be 1000 and `pickWithProb()` would return `false`. It *think* that `pick(0, 999)` would fix this problem and be correct for all inputs.

Reply:
Good catch, you're right, this should be `pick(0, 999)`.